src/access/Subjects.h

```diff
@@ -0,0 +1,208 @@
+#include <lib/support/Variant.h>
+
+namespace chip {
+namespace Access {
+
+struct PaseSubject
+{
+    static constexpr uint16_t kPasscodeId = 0;
+    bool operator==(const PaseSubject & that) const { return true; }
+};
+
+struct NodeSubject
+{
+    NodeSubject(NodeId aNodeId) : nodeId(aNodeId) {}
+    NodeId nodeId;
+    bool operator==(const NodeSubject & that) const { return this->nodeId == that.nodeId; }
+};
+
+struct GroupSubject
+{
+    GroupSubject(GroupId aGroupId) : groupId(aGroupId) {}
+    GroupId groupId;
+    bool operator==(const GroupSubject & that) const { return this->groupId == that.groupId; }
+};
+
+class OperationalNodeId;
+
+/** A scoped subject is a unique identifier with-in a fabric scope. */
+class ScopedSubject
+{
+public:
+    ScopedSubject() {}
+
+    template <typename T, class... Args>
+    constexpr explicit ScopedSubject(InPlaceTemplateType<T>, Args &&... args) :
+        mSubject(InPlaceTemplate<T>, std::forward<Args>(args)...)
+    {}
+
+    template <typename T, typename... Args>
+    static ScopedSubject Create(Args &&... args)
+    {
+        return ScopedSubject(InPlaceTemplate<T>, std::forward<Args>(args)...);
+    }
+
+    Access::AuthMode GetAuthMode() const
+    {
+        if (mSubject.Is<PaseSubject>())
+        {
+            return Access::AuthMode::kPase;
+        }
+        else if (mSubject.Is<NodeSubject>())
+        {
+            return Access::AuthMode::kCase;
+        }
+        else if (mSubject.Is<GroupSubject>())
+        {
+            return Access::AuthMode::kGroup;
+        }
+        else
+        {
+            return Access::AuthMode::kNone;
+        }
+    }
+
+#if CHIP_DETAIL_LOGGING
+    const char * GetAuthModeString() const
+    {
+        switch (GetAuthMode())
+        {
+        case Access::AuthMode::kPase:
+            return "Pase";
+        case Access::AuthMode::kCase:
+            return "Case";
+        case Access::AuthMode::kGroup:
+            return "Group";
+        case Access::AuthMode::kNone:
+        default:
+            return "None";
+        }
+    }
+#endif // CHIP_DETAIL_LOGGING
+
+    /** Return subject value described in spec */
+    uint64_t GetValue() const
+    {
+        if (mSubject.Is<PaseSubject>())
+        {
+            return static_cast<uint64_t>(mSubject.Get<PaseSubject>().kPasscodeId) << 48;
+        }
+        else if (mSubject.Is<NodeSubject>())
+        {
+            return mSubject.Get<NodeSubject>().nodeId;
+        }
+        else if (mSubject.Is<GroupSubject>())
+        {
+            return static_cast<uint64_t>(mSubject.Get<GroupSubject>().groupId) << 48;
+        }
+        else
+        {
+            return std::numeric_limits<uint64_t>::max();
+        }
+    }
+
+    bool operator==(const ScopedSubject & that) const { return this->mSubject == that.mSubject; }
+
+private:
+    Variant<PaseSubject, NodeSubject, GroupSubject> mSubject;
+};
+
+/**
+ * A subject is a global unique identifier. It suite 2 purpose:
```

Comment:
A subject as defined here is very much not a "global unique identifier".  It's a unique subject identifier within the context of a given node, but that's all it can be, given how fabric indices work.

Reply:
Thanks. I'll update the comment